dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,734 @@
-  anchor_out_data[1] = coordinates[1];
+The order of dimensions is determined by the layout provided in ``shape_layout``.
 
-  // Copy the offsets to output 1
-  auto &offsets_out = ws.Output<CPUBackend>(1);
-  offsets_out.Resize({2});
+Note: ``crop_shape`` and ``input_shape`` should be provided together, and providing those is
+incompatible with using ``scaling`` and ``aspect_ratio`` arguments.)code",
+                         std::vector<int>{}, true)
+    .AddOptionalArg<TensorLayout>(
+        "bbox_layout",
+        R"code(Determines the meaning of the coordinates of the bounding boxes.
+  Possible values are:
 
-  auto *offsets_out_data = offsets_out.mutable_data<float>();
-  offsets_out_data[0] = coordinates[2];
-  offsets_out_data[1] = coordinates[3];
-}
+  ``x`` (horizontal start anchor), ``y`` (vertical start anchor), ``z`` (depthwise start anchor),
 
-template <>
-void RandomBBoxCrop<CPUBackend>::WriteBoxesToOutput(
-    SampleWorkspace &ws, const BoundingBoxes &bounding_boxes) const {
-  auto &bbox_out = ws.Output<CPUBackend>(2);
-  bbox_out.Resize(
-      {static_cast<int64_t>(bounding_boxes.size()), static_cast<int64_t>(BoundingBox::kSize)});
-
-  auto *bbox_out_data = bbox_out.mutable_data<float>();
-  for (size_t i = 0; i < bounding_boxes.size(); ++i) {
-    auto *output = bbox_out_data + i * BoundingBox::kSize;
-    auto coordinates =
-        ltrb_ ? bounding_boxes[i].AsLtrb() : bounding_boxes[i].AsXywh();
-    output[0] = coordinates[0];
-    output[1] = coordinates[1];
-    output[2] = coordinates[2];
-    output[3] = coordinates[3];
+  ``X`` (horizontal end anchor),   ``Y`` (vertical end anchor),   ``Z`` (depthwise end anchor),
+
+  ``W`` (width),                   ``H`` (height),                ``D`` (depth).
+
+Note: If left empty, ``"xyXY"`` or ``"xyzXYZ"`` will be assumed, depending on the number of dimensions.
+)code",
+        TensorLayout{""})
+    .AddOptionalArg<TensorLayout>(
+        "shape_layout",
+        R"code(Determines the meaning of the dimensions provided in `crop_shape` and `input_shape`.
+Possible values are:
+
+  ``W`` (width), ``H`` (height), ``D`` (depth).
+
+Note: If left empty, ``"WH"`` or ``"WHD"`` will be assumed, depending on the number of dimensions.
+)code",
+        TensorLayout{""});
+
+template <int ndim>
+class RandomBBoxCropImpl : public detail::OpImplBase<CPUBackend> {
+ public:
+  static constexpr int coords_size = ndim * 2;
+
+  ~RandomBBoxCropImpl() = default;
+
+  explicit RandomBBoxCropImpl(const OpSpec &spec)
+      : spec_(spec)
+      , num_attempts_{spec.GetArgument<int>("num_attempts")}
+      , has_labels_(spec.NumInput() > 1)
+      , has_crop_shape_(spec.ArgumentDefined("crop_shape"))
+      , has_input_shape_(spec.ArgumentDefined("input_shape"))
+      , bbox_layout_(spec.GetArgument<TensorLayout>("bbox_layout"))
+      , shape_layout_(spec.GetArgument<TensorLayout>("shape_layout"))
+      , rngs_(spec.GetArgument<int64_t>("seed"), spec.GetArgument<int>("batch_size")) {
```

Comment:
Isn't this formatting incompliant with [Google style](https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists)?

Reply:
I found this way nicer to read, as the commas are aligned with the `:`. Don't really care much though. If you insist I can change it

Reply:
Changed it